### DISCUSSION TITLE
How to do visual debugging of a PDF

### DISCUSSION THREAD

**Discussion by mlissner** [pymupdf/PyMuPDF#1260] [Looking for help] [answered]
I saw in a post a few days ago (sorry, I lost it), a PDF that had all of the elements outlined in different colors so you could see the edges of each piece of text. 

Is there a trick for doing that? I don't think I've seen it in the docs, but it would help with a project I'm doing related to the intersection of rectangles with words.

Thank you!

**Comment by JorjMcKie** [comment]
did you mean this: https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/layout-demo1.pdf

**Reply by mlissner** [reply]
I actually just found it. It's this, but same idea:

https://github.com/pymupdf/PyMuPDF/issues/318#issuecomment-657095999

The actual rectangle I'm trying to understand is the first redaction in the attached:

![Screenshot from 2021-09-13 13-38-55](https://user-images.githubusercontent.com/236970/133153090-3e61ca7d-14e2-447b-8ce6-47cdb66ad128.png)

I think the rectangle's bbox spans all three lines and thus includes the text you can see in the screenshot, but I can't be sure. For example, the text on the first and third lines is getting flagged as part of the redaction when it's not under the black part of the rectangle.

Here's the doc if you're curious:

[partial_intersection_no_bad_redaction_no_hits.pdf](https://github.com/pymupdf/PyMuPDF/files/7157230/partial_intersection_no_bad_redaction_no_hits.pdf)

**Reply by JorjMcKie** [reply]
This and similar examples are handled in folder https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/word%26line-marking.

The arked text here has probably been found via text search. The returned single rectangles of this method jointly make up the marked area. The highlight annotation's rect (if it actually is a highight annot) is the union of these rectangles and in this case would cover all three complete lines.

**Reply by mlissner** [reply]
Thanks. That wasn't quite what I was looking for — I'm trying to find text under bad redactions — but I did find this, which sort of helped: https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/layout-analyzer.py. It gave me some of the outlines of the boxes that are in the PDF, but I'm still trying to figure out how to do this properly. I'll keep digging, thank you.

**Comment by mlissner** [comment, marked as answer]
In the end, the issue I had was that I was doing:

```
drawings = some_page.get_drawings()
```

And then:
```
rect = drawings[0].rect
```

That totally doesn't work for multi-line redactions like I had. In those cases, you have to read the docs, which explain that the `drawing["items"]` key is where the real shapes are. Once I started using those, I was able to handle the multi-line redactions no problem.

**Comment by JorjMcKie** [comment]
> you have to read the docs

I could not agree more 😎👍😉
